clang/runtime/dpct-rt/include/dpct/dpl_extras/algorithm.h

```diff
@@ -978,7 +1018,41 @@ stable_partition(Policy &&policy, Iter1 first, Iter1 last, Iter2 mask, Pred p) {
 }
 
 template <typename Policy, typename Iter1, typename Iter2, typename Pred>
-internal::enable_if_execution_policy<Policy, Iter1>
+typename std::enable_if<internal::is_hetero_execution_policy<
+                            typename std::decay<Policy>::type>::value,
+                        Iter1>::type
+partition(Policy &&policy, Iter1 first, Iter1 last, Iter2 mask, Pred p) {
+  static_assert(
+      std::is_same<typename std::iterator_traits<Iter1>::iterator_category,
+                   std::random_access_iterator_tag>::value &&
+          std::is_same<typename std::iterator_traits<Iter2>::iterator_category,
+                       std::random_access_iterator_tag>::value,
+      "Iterators passed to algorithms must be random-access iterators.");
+  auto _partition_call1 =
+      internal::make_wrapped_policy<internal::partition_call1>(policy);
+  auto _copy_call1 = internal::make_wrapped_policy<internal::copy_call1>(
+      std::forward<Policy>(policy));
+
+  auto _n = std::distance(first, last);
+  internal::__buffer<typename std::iterator_traits<Iter1>::value_type> _tmp(_n);
+
+  auto _tmp_first = _tmp.get();
+  auto _tmp_last = _tmp_first + _n;
+  auto _tmp_last_rev = std::reverse_iterator(_tmp_last);
+
+  auto _end_pair =
+      stable_partition_copy(std::move(_partition_call1), first, last, mask,
+                            _tmp_first, _tmp_last_rev, p);
+  auto _first_n = std::distance(_tmp_first, std::get<0>(_end_pair));
+  std::copy(std::move(_copy_call1), _tmp_first, _tmp_last, first);
```

Comment:
Same as above.

Reply:
I don't think this is possible in this one, because we only have one copy call, which covers both parts of the partitioned result.  (This is what breaks the stability, because we are copying in a reversed second part).

I may be missing something, but I think we need the distance call and addition in this version.